core/metacling/src/TCling.cxx

```diff
@@ -1779,29 +1792,21 @@ void TCling::RegisterModule(const char* modulename,
       // its symbols are not yet reachable from the process.
       // Recursive dlopen seems to work just fine.
       dyLibName = FindLibraryName(triggerFunc);
+
       if (dyLibName) {
-         // We were able to determine the library name.
+         // The triggerFunc may not be in a shared object but in an executable.
+         // In that case, we are done.
+         if (!cling::DynamicLibraryManager::isSharedLibrary(dyLibName))
+            return;
+
+         // We were able to determine that is a valid library.
          void* dyLibHandle = dlopen(dyLibName, RTLD_LAZY | RTLD_GLOBAL);
-         if (!dyLibHandle) {
-#ifdef R__WIN32
-            char dyLibError[1000];
-            FormatMessageA(FORMAT_MESSAGE_FROM_SYSTEM, NULL, GetLastError(), MAKELANGID(LANG_NEUTRAL, SUBLANG_DEFAULT),
-                           dyLibError, sizeof(dyLibError), NULL);
-#else
-            const char* dyLibError = dlerror();
-            if (dyLibError)
-#endif
-            {
-               if (gDebug > 0) {
-                  ::Info("TCling::RegisterModule",
-                         "Cannot open shared library %s for dictionary %s:\n  %s",
-                         dyLibName, modulename, dyLibError);
-               }
-            }
-            dyLibName = 0;
-         } else {
+         if (dyLibHandle) {
             fRegisterModuleDyLibs.push_back(dyLibHandle);
-         } // if (!dyLibHandle) .. else
+         } else {
+            MaybePrintDlError(dyLibName, modulename);
+            return;
```

Comment:
The `dlopen` happens (as the comment nicely reminds me) for resolving symbols into that just-opened shared library. To be able to resolve them, we need to *recursively* `dlopen` it *again*.

If the `dlopen` fails (e.g. because it is in fact *not* a shared library) then that's not a huge problem: if the JIT then fails to resolve symbols then so be it, but that's waaaay more unlikely than things just working out fine.

We should at least try to initialize the dictionary (say in the binary) instead of returning here. I.e. I believe this code must survive `dyLibHandle == nullptr`

Reply:
Do you remember which symbols you had to resolve? I suspect something w.r.t. the static initialization itself?

Reply:
Yes, if the interpreter is used during static init.